pkg/cfn/manager/tasks.go

```diff
@@ -1,158 +1,215 @@
+		logger.Debug("no actual tasks")
+		return nil
+	}
+
+	errs := make(chan error)
 
-// DeleteAllNodeGroups deletes all nodegroups without waiting
-func (c *StackCollection) DeleteAllNodeGroups(call taskFunc) []error {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return []error{err}
+	if t.Parallel {
+		go doParallelTasks(errs, t.tasks)
+	} else {
+		go doSequentialTasks(errs, t.tasks)
 	}
 
-	errs := []error{}
-	for _, s := range nodeGroupStacks {
-		if err := c.DeleteNodeGroup(getNodeGroupName(s)); err != nil {
-			errs = append(errs, err)
-		}
+	allErrs := []error{}
+	for err := range errs {
+		allErrs = append(allErrs, err)
 	}
+	return allErrs
+}
 
-	return errs
+type taskWithoutParams struct {
+	info string
+	call func(chan error) error
 }
 
-// WaitDeleteAllNodeGroups runs all tasks required to delete all the nodegroup
-// stacks and wait for all nodegroups to be deleted; any errors will be returned
-// as a slice as soon as the group of tasks is completed
-func (c *StackCollection) WaitDeleteAllNodeGroups(force bool) []error {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return []error{err}
-	}
+func (t *taskWithoutParams) Describe() string { return t.info }
+func (t *taskWithoutParams) Do(errs chan error) error {
+	return t.call(errs)
+}
 
-	errs := []error{}
-	appendErr := func(err error) {
-		errs = append(errs, err)
-	}
+type taskWithNameParam struct {
+	info string
+	name string
+	call func(chan error, string) error
+}
 
-	deleteAllNodeGroups := []Task{}
-	for i := range nodeGroupStacks {
-		t := Task{
-			Call: c.WaitDeleteNodeGroup,
-			Data: getNodeGroupName(nodeGroupStacks[i]),
-		}
-		if force {
-			t.Call = c.WaitForceDeleteNodeGroup
-		}
-		deleteAllNodeGroups = append(deleteAllNodeGroups, t)
+func (t *taskWithNameParam) Describe() string { return t.info }
+func (t *taskWithNameParam) Do(errs chan error) error {
+	return t.call(errs, t.name)
+}
+
+type taskWithNodeGroupSpec struct {
+	info      string
+	nodeGroup *api.NodeGroup
+	call      func(chan error, *api.NodeGroup) error
+}
+
+func (t *taskWithNodeGroupSpec) Describe() string { return t.info }
+func (t *taskWithNodeGroupSpec) Do(errs chan error) error {
+	return t.call(errs, t.nodeGroup)
+}
+
+type taskWithStackSpec struct {
+	info  string
+	stack *Stack
+	call  func(*Stack, chan error) error
+}
+
+func (t *taskWithStackSpec) Describe() string { return t.info }
+func (t *taskWithStackSpec) Do(errs chan error) error {
+	return t.call(t.stack, errs)
+}
+
+type asyncTaskWithStackSpec struct {
+	info  string
+	stack *Stack
+	call  func(*Stack) (*Stack, error)
+}
+
+func (t *asyncTaskWithStackSpec) Describe() string { return t.info + " [async]" }
+func (t *asyncTaskWithStackSpec) Do(errs chan error) error {
+	_, err := t.call(t.stack)
+	close(errs)
+	return err
+}
+
+func doSingleTask(allErrs chan error, task Task) {
+	desc := task.Describe()
+	logger.Debug("started task: %s", desc)
+	errs := make(chan error)
```

Comment:
Why not share the same allErrs channel?

Reply:
defer errs.close()?

Reply:
> Why not share the same allErrs channel?

The task itself is expected to close this channel, and what we see on this channel gets copied into shared `allErrs` channel. I don't think we should change that now.